### PR TITLE
Date time2

### PR DESCRIPTION
**Pull request by armanvatandoust**
* [ ] Same coding style as existing code
* [ ] Variable naming, comment style, documentation
* [ ] Make sure the Method decorator has the correct attributes
* [ ] Check all user inputted values’ types
* [ ] Protect against “bad” inputs
* [ ] Both kwargs and args should work
* [ ] Prefer Python types
* [ ] Avoid hard-coded (magic) strings/numbers
* [ ] Make sure raised exceptions have the same string representation as the ones in Python
* [ ] Make sure flake8 & ant checkstyle pass without errors

### CODE REVIEW

File: python/common/org/python/stdlib/datetime/Date.java [hunk 1]
```diff
@@ -1,32 +1,42 @@
...
+import java.util.Arrays;
+import java.util.Calendar;
+import static java.lang.Math.toIntExact;
+
+import org.python.types.Bool;
+
+public class Date extends org.python.types.Object {
+    private Int year;
+    private Int month;
+    private Int day;
```
Comment (Huntaz) [inline python/common/org/python/stdlib/datetime/Date.java]: Modify Int toorg.python.types.Int everywhere to keep in line to the same coding style as the rest of the project.

Comment (Huntaz) [inline python/common/org/python/stdlib/datetime/Date.java]: Modify `Int` to `org.python.types.Int `everywhere to keep in line to the same coding style as the rest of the project.

Comment (Huntaz) [reply python/common/org/python/stdlib/datetime/Date.java]: realized this is Date... but still

File: python/common/org/python/stdlib/datetime/Date.java [hunk 2]
```diff
@@ -1,32 +1,42 @@
...
+import org.python.types.Map;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Calendar;
+import static java.lang.Math.toIntExact;
+
+import org.python.types.Bool;
+
+public class Date extends org.python.types.Object {
```
Comment (Huntaz) [inline python/common/org/python/stdlib/datetime/Date.java]: Remove Date from your pull request

File: python/common/org/python/stdlib/datetime/DateTime.java [hunk 3]
```diff
@@ -48,10 +48,9 @@
     private ArrayList daysInMonth;
 
     @org.python.Method(
-        __doc__ = "TODO: DateTime",
+        __doc__ = "DateTime constructor",
         default_args = {}
```
Comment (marcusfroling) [inline python/common/org/python/stdlib/datetime/DateTime.java]: Empty default_args not necessary

File: tests/stdlib/test_datetime.py [hunk 4]
```diff
@@ -175,6 +175,16 @@ def test_DateTime_with_to_many_args(self):
             """)
 
 
+    def test_DateTime_replace_to_many_args(self):
+        self.assertCodeExecution("""
+            from datetime import datetime
+            try:
+                d = datetime(1,1,1,1,1,1,1)
```
Comment (marcusfroling) [inline tests/stdlib/test_datetime.py]: Should be spacing after comma.
I.e. (1,1,1) should be (1, 1, 1)

File: tests/stdlib/test_datetime.py [hunk 5]
```diff
@@ -465,7 +493,7 @@ def test_Date_today_valid_input(self):
                 print(e)
             """)
 
-    @expectedFailure
+    # @expectedFailure
```
Comment (marcusfroling) [inline tests/stdlib/test_datetime.py]: Unnecessary comment

File: tests/stdlib/test_datetime.py [hunk 6]
```diff
@@ -260,7 +260,7 @@ def test_DateTime_replace_microsecond(self):
             except TypeError as err:
                 print(err)
             """)
-    def test_Date(self):
+    def test_DateTime_Date(self):
```
Comment (marcusfroling) [inline tests/stdlib/test_datetime.py]: DateTime should be lowercase to match rest of project

File: python/common/org/python/stdlib/datetime/DateTime.java [hunk 7]
```diff
@@ -0,0 +1,442 @@
...
+    }
+
+    @org.python.Method(
+            __doc__ = "Enables either month list with leap year or without",
+            args = { "year" }
+    )
+    private void setDaysInMonth(org.python.types.Int year) {
+        if (isLeapYear(year)) {
+            this.daysInMonth = new ArrayList<Integer>(Arrays.asList(-1, 31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31));
+        } else {
```
Comment (Huntaz) [inline python/common/org/python/stdlib/datetime/DateTime.java]: I, for some reason, recognize this part

File: python/common/org/python/stdlib/datetime/DateTime.java [hunk 8]
```diff
@@ -0,0 +1,442 @@
...
+        return new Boolean((day.value <= (Integer) this.daysInMonth.get(toIntExact(month.value))) && (day.value > 0));
+    }
+
+    @org.python.Method(
+            __doc__ = "Checks the inputs given is correct"
+
+    )
+    private void checkIfInputDateIsValid(org.python.types.Int year, org.python.types.Int month, org.python.types.Int day, org.python.types.Int hour,
+                                         org.python.types.Int minute, org.python.types.Int second, org.python.types.Int microsecond) {
+        if (year.value < 1 || year.value > 10000) {
```
Comment (Huntaz) [inline python/common/org/python/stdlib/datetime/DateTime.java]: you have MAXYEAR(?)

Comment (Huntaz) [reply python/common/org/python/stdlib/datetime/DateTime.java]: And MINYEAR

File: python/common/org/python/stdlib/datetime/DateTime.java [hunk 9]
```diff
@@ -0,0 +1,442 @@
...
+        setDaysInMonth(year);
+        if (!isDayInMonth(month, day)) {
+            throw new org.python.exceptions.ValueError("day is out of range for month");
+        }
+    }
+
+    @org.python.Method(
+            __doc__ = "Creates a Date object from the DateTime object"
+    )
+
```
Comment (Huntaz) [inline python/common/org/python/stdlib/datetime/DateTime.java]: remove linebreak according to style guide

File: python/common/org/python/stdlib/datetime/DateTime.java [hunk 10]
```diff
@@ -0,0 +1,442 @@
...
+    )
+    private String filler(String input, int target) {
+        int len = input.toString().length();
+        String output = input;
+        for (int i = len; i < target; i++) {
+            output = "0" + output;
+        }
+        return output;
+    }
+    @org.python.Method(
```
Comment (Huntaz) [inline python/common/org/python/stdlib/datetime/DateTime.java]: include line break according to style

File: python/common/org/python/stdlib/datetime/DateTime.java [hunk 11]
```diff
@@ -0,0 +1,442 @@
...
+            result += "." + microsecond;
+        }
+        return new org.python.types.Str(result);
+    }
+
+    @org.python.Method(
+            __doc__ = "Returns a DateTime object with the current date and time"
+    )
+    public static DateTime now() {
+        LocalDateTime now = LocalDateTime.now();
```
Comment (Huntaz) [inline python/common/org/python/stdlib/datetime/DateTime.java]: neat solution

File: python/common/org/python/stdlib/datetime/DateTime.java [hunk 12]
```diff
@@ -241,6 +241,10 @@ private Boolean isDayInMonth(org.python.types.Int month, org.python.types.Int da
     private void checkIfInputDateIsValid(org.python.types.Int year, org.python.types.Int month, org.python.types.Int day, org.python.types.Int hour,
                                          org.python.types.Int minute, org.python.types.Int second, org.python.types.Int microsecond)
     {
+        if(year.value < 1 || year.value > 10000){
```
Comment (marcusfroling) [inline python/common/org/python/stdlib/datetime/DateTime.java]: MAX_YEAR is 9999, but here you include 10000.

File: python/common/org/python/stdlib/datetime/DateTime.java [hunk 13]
```diff
@@ -0,0 +1,442 @@
...
+        Int tmpHour = this.hour;
+        Int tmpMinute = this.minute;
+        Int tmpSecond = this.second;
+        Int tmpMicroSecond = this.microsecond;
+        Str tmpTzinfo = this.tzinfo;
+
+        if (year instanceof org.python.types.Int) {
+            tmpYear = (Int) year;
+        } else if (year != null) {
+            String typeName = year.getClass().getSimpleName().toLowerCase();
```
Comment (Huntaz) [inline python/common/org/python/stdlib/datetime/DateTime.java]: use typeName() if possible to get class name

File: python/common/org/python/stdlib/datetime/DateTime.java [hunk 14]
```diff
@@ -0,0 +1,442 @@
...
+
+        int argsTotalLength = args.length + kwargs.size();
+
+        if ((argsTotalLength) > 8) {
+            throw new org.python.exceptions.TypeError("Function takes at most 8 positional arguments (" + (argsTotalLength) + " given)");
+        }
+
+        org.python.Object yearKwargs = kwargs.get("year");
+        if (args.length >= 1) {
+            year = (Int) args[0];
```
[the remark this answers is not in the document]
Comment (Druvan) [reply python/common/org/python/stdlib/datetime/DateTime.java]: We are checking that on the lines after

File: python/common/org/python/stdlib/datetime/Date.java (hunk 2 above)
Comment (Druvan) [reply python/common/org/python/stdlib/datetime/Date.java]: I will merge with your branch before mege into "master" so we have the same code in Date class

File: python/common/org/python/stdlib/datetime/DateTime.java [hunk 15]
```diff
@@ -0,0 +1,442 @@
...
+
+        checkIfInputDateIsValid(year, month, day, hour, minute, second, microsecond);
+
+    }
+
+    @org.python.Method(
+            __doc__ = "DateTime java constructor",
+            args = { "year", "month", "day", "hour", "minute", "second", "microsecond", "tzinfo" }
+    )
+    public DateTime(int year, int month, int day, int hour, int minute, int second, int microsecond, String tzinfo) {
```
[the remark this answers is not in the document]
Comment (Druvan) [reply python/common/org/python/stdlib/datetime/DateTime.java]: Changed to protected

**Comment by Huntaz** [comment]
My alias is Hunter and I approve of this Pull Request

File: tests/stdlib/test_datetime.py (hunk 5 above)
Comment (marcusfroling) [reply tests/stdlib/test_datetime.py]: Haha indeed it is.

**Pull request merged**